packages/block-editor/src/hooks/margin.js

```diff
@@ -124,26 +125,13 @@ export function MarginEdit( props ) {
 		} );
 	};
 
-	const onChangeShowVisualizer = ( next ) => {
-		const newStyle = {
-			...style,
-			visualizers: {
-				margin: next,
-			},
-		};
-
```

Comment:
Could we just use `__unstableMarkNextChangeAsNotPersistent` here and `__unstableMarkLastChangeAsPersistent` before the `setAttributes` at the `onChange` function? The second action is because of this: https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/store/reducer.js#L497

>All changes are considered persistent except when updating the same block attribute as in the previous action.

Reply:
> Could we just use `__unstableMarkNextChangeAsNotPersistent` here and `__unstableMarkLastChangeAsPersistent` before the `setAttributes` at the `onChange` function?

I wasn't aware of these, but after skimming through the code, if I understand it correctly, it looks like they are basically just a complicated way to toggle between triggering the `onInput` versus `onChange` props for the `BlockEditor`. It doesn't really solve the root-cause of this issue—using `attributes` to store editor-specific data.

> I was wondering if the alternative I propose in the comment could be considered and I'm not sure if it is too hacky 😄 .

Using `attributes` for anything other than properties needed to render the final block is what I think is the hacky part. As long as we're using `attributes` for editor data, there are always going to be opportunities for bugs like this one to crop up.

> This way we could add `BoxControlVisualizer` to other blocks that use `padding` support without any other change needed.

The ideal case would be incorporating `BoxControlVisualizer` directly into the hook so only the block.json needs to be updated for the support. Maybe someone has an idea for how to do that in the overhaul that @aaronrobertshaw mentioned. The changes for blocks using the hook are pretty minimal with this approach, but you're right, they aren't nothing.

Reply:
>The ideal case would be incorporating BoxControlVisualizer directly into the hook so only the block.json needs to be updated for the support. 

Exactly!

Also I wasn't able to reproduce adding extra props in `style` besides the `padding ones`. Do I have to do something specific?

Reply:
> The ideal case would be incorporating BoxControlVisualizer directly into the hook so only the block.json needs to be updated for the support. Maybe someone has an idea for how to do that in the overhaul that @aaronrobertshaw mentioned. 

This might not be so straightforward when the differences between blocks (and their markup) are taken into account. More complicated again, when both padding and margin wish to be visualized. 

With the current visualizer, displaying margins is particularly problematic and can require adding a wrapping element which creates a disjoint between the markup in the editor and the frontend. 

One previous suggestion on how the visualizer could work was making it not rely on a parent element and instead maybe use a ref/popover.

For the moment, I'd think being able to handle the visualization on a block by block basis would be best.